internal/session/sessionwantsender.go

```diff
@@ -354,21 +361,59 @@ func (spm *sessionWantSender) processUpdates(updates []update) {
 		}
 	}
 
-	// If all available peers for a cid sent a DONT_HAVE, signal to the session
-	// that we've exhausted available peers
-	if dontHaves.Len() > 0 {
-		exhausted := spm.bpm.AllPeersDoNotHaveBlock(spm.peerAvlMgr.availablePeers(), dontHaves.Keys())
-		newlyExhausted := spm.newlyExhausted(exhausted)
-		if len(newlyExhausted) > 0 {
-			spm.onPeersExhausted(newlyExhausted)
-		}
-	}
-
 	// If any peers have sent us too many consecutive DONT_HAVEs, remove them
 	// from the session
 	for p := range prunePeers {
 		spm.SignalAvailability(p, false)
 	}
+
+	return dontHaves
+}
+
+// checkForExhaustedWants checks if there are any wants for which all peers
+// have sent a DONT_HAVE. We call these "exhausted" wants.
+func (spm *sessionWantSender) checkForExhaustedWants(dontHaves *cid.Set, newlyUnavailable []peer.ID) {
+	// If there are no new DONT_HAVEs, and no peers became unavailable, then
+	// we don't need to check for exhausted wants
+	if dontHaves.Len() == 0 && len(newlyUnavailable) == 0 {
+		return
+	}
+
+	// We need to check each want for which we just received a DONT_HAVE
+	wants := dontHaves.Keys()
+
+	// If a peer just became unavailable, then we need to check all wants
+	// (because it may be the last peer who hadn't sent a DONT_HAVE for a CID)
+	if len(newlyUnavailable) > 0 {
+		// Collect all pending wants
+		wants = make([]cid.Cid, len(spm.wants))
```

Comment:
We just set this above, should we be overwriting it?

Reply:
The block presence manager has a table of which peers have send a HAVE / DONT_HAVE for each want (CID).

So when we receive a DONT_HAVE, we need to check against the table to see if all peers have now sent us a DONT_HAVE for that want (CID).

In the case where a peer has become unavailable, we need to check against all wants (not just the DONT_HAVEs we just received) because it may be that the peer was the last peer who we were waiting for a response from.

Reply:
Sure, but what if we have both new `dontHaves` _and_ `newlyUnavailable` peers (specifically, from _different_ peers)?

Reply:
Oh, nevermind. `wants` is a superset.